broker/src/bin/neon_broker.rs

```diff
@@ -0,0 +1,563 @@
+use tracing::*;
+
+use neon_broker::neon_broker_proto::neon_broker_server::{NeonBroker, NeonBrokerServer};
+use neon_broker::neon_broker_proto::subscribe_safekeeper_info_request::SubscriptionKey as ProtoSubscriptionKey;
+use neon_broker::neon_broker_proto::{SafekeeperTimelineInfo, SubscribeSafekeeperInfoRequest};
+use utils::id::TenantTimelineId;
+use utils::logging::{self, LogFormat};
+use utils::project_git_version;
+
+project_git_version!(GIT_VERSION);
+
+#[derive(Parser, Debug)]
+#[command(version = GIT_VERSION, about = "Broker for neon storage nodes communication", long_about = None)]
+struct Args {
+    /// Endpoint to listen on.
+    #[arg(short, long, default_value = DEFAULT_LISTEN_ADDR)]
+    listen_addr: SocketAddr,
+    /// Size of the queue to the subscriber.
+    #[arg(long, default_value_t = 256)]
+    chan_size: usize,
+    /// HTTP/2 keepalive interval.
+    #[arg(long, value_parser= humantime::parse_duration, default_value = "5000ms")]
+    http2_keepalive_interval: Duration,
+    /// Format for logging, either 'plain' or 'json'.
+    #[arg(long, default_value = "plain")]
+    log_format: String,
+}
+
+// Args is put into OnceCell to make conf available everywhere.
+static ARGS: OnceCell<Args> = OnceCell::new();
+fn get_args() -> &'static Args {
+    ARGS.get().expect("ARGS is not initialized")
+}
+
+type PubId = u64; // id of publisher for registering in maps
+type SubId = u64; // id of subscriber for registering in maps
+
+#[derive(Copy, Clone)]
+enum SubscriptionKey {
+    All,
+    Timeline(TenantTimelineId),
+}
+
+impl SubscriptionKey {
+    // Parse protobuf subkey (protobuf doesn't have fixed size bytes, we get vectors).
+    pub fn from_proto_subscription_key(key: ProtoSubscriptionKey) -> Result<Self, Status> {
+        match key {
+            ProtoSubscriptionKey::All(_) => Ok(SubscriptionKey::All),
+            ProtoSubscriptionKey::TenantTimelineId(proto_ttid) => {
+                Ok(SubscriptionKey::Timeline(parse_proto_ttid(&proto_ttid)?))
+            }
+        }
+    }
+}
+
+// Subscriber id + tx end of the channel for messages to it.
+#[derive(Clone)]
+struct SubSender(SubId, Sender<SafekeeperTimelineInfo>);
+
+impl fmt::Debug for SubSender {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Subscription id {}", self.0)
+    }
+}
+
+// Announcements subscriber sends to publisher(s) asking it to stream to the
+// provided channel, or forget about it, releasing memory.
+#[derive(Clone)]
+enum SubAnnounce {
+    AddAll(Sender<SafekeeperTimelineInfo>), // add subscription to all timelines
+    AddTimeline(TenantTimelineId, SubSender), // add subsciption to the specific timeline
+    RemoveTimeline(TenantTimelineId, SubId), // remove subscription to the specific timeline
+                                            // RemoveAll is not needed as publisher will notice closed channel while
+                                            // trying to send the next message.
```

Comment:
Since this is new code a nit: would prefer to see rustdoc friendly comments before the variants.

```suggestion
    /// Add subscription to all timelines
    AddAll(Sender<SafekeeperTimelineInfo>),
    /// Add subscription to the specific timeline
    AddTimeline(TenantTimelineId, SubSender),
    /// Remove subscription to the specific timeline
    RemoveTimeline(TenantTimelineId, SubId),
    
    // RemoveAll is not needed as publisher will notice closed channel while
    // trying to send the next message.
```